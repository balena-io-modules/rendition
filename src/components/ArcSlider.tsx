import * as React from 'react';
import { ArcSliderProps } from 'rendition';
import styled from 'styled-components';
import { constrainNumber, darken, getAngleBetweenPoints } from '../utils';
import { Box, Flex } from './Grid';

const ANGLERANGE = 135;

const StyledSVG = styled.svg`
	display: block;
	width: 100%;
	height: 100%;
`;

const ArcSliderContainer = styled(Box)`
	position: relative;
`;

const LabelContainer = styled(Flex).attrs({
	justify: 'center',
	flexDirection: 'column',
})`
	position: absolute;
	top: 0;
	right: 0;
	bottom: 0;
	left: 0;
	text-align: center;
	pointer-events: none;
`;

const polarToCartesian = (
	centerX: number,
	centerY: number,
	radius: number,
	angleInDegrees: number,
) => {
	const angleInRadians = (angleInDegrees - 90) * Math.PI / 180.0;

	return {
		x: centerX + radius * Math.cos(angleInRadians),
		y: centerY + radius * Math.sin(angleInRadians),
	};
};

const describeArc = (
	x: number,
	y: number,
	radius: number,
	startAngle: number,
	endAngle: number,
) => {
	const start = polarToCartesian(x, y, radius, endAngle);
	const end = polarToCartesian(x, y, radius, startAngle);
	const largeArcFlag = endAngle - startAngle <= 180 ? '0' : '1';

	const d = `M ${start.x} ${
		start.y
	} A ${radius} ${radius} 0 ${largeArcFlag} 0 ${end.x} ${end.y}`;

	return d;
};

interface ArcSliderState {
	isDragging: boolean;
	innerValue: number;
}

class ArcSlider extends React.Component<ArcSliderProps, ArcSliderState> {
	private $slider: SVGSVGElement | null = null;

	constructor(props: any) {
		super(props);

		const { value } = props;

		this.state = {
			isDragging: false,
			innerValue: value == null ? 0 : constrainNumber(value, 0, 1),
		};
	}

	componentDidMount() {
		document.addEventListener('mousemove', this.handleMouseMove);
		document.addEventListener('touchmove', this.handleMouseMove);
		document.addEventListener('mouseup', this.handleMouseUp);
		document.addEventListener('touchend', this.handleMouseUp);
		document.addEventListener('touchcancel', this.handleMouseUp);
	}

	componentWillUnmount() {
		document.removeEventListener('mousemove', this.handleMouseMove);
		document.removeEventListener('touchmove', this.handleMouseMove);
		document.removeEventListener('mouseup', this.handleMouseUp);
		document.removeEventListener('touchend', this.handleMouseUp);
		document.removeEventListener('touchcancel', this.handleMouseUp);
	}

	handleMouseUp = () => {
		this.setState({ isDragging: false });
	};

	handleMouseDown = (e: any) => {
		e.preventDefault();
		this.setState({ isDragging: true });
	};

	handleMouseMove = (e: any) => {
		if (this.state.isDragging && this.$slider) {
			const { left, top, width, height } = this.$slider.getBoundingClientRect();

			const sliderCenter = {
				x: left + width / 2,
				y: top + height / 2,
			};

			let mouseLocation = {
				x: e.pageX,
				y: e.pageY,
			};

			// If this is a touch event, the page coordinates need to be retrieved
			// from the `touches` collection
			if (!!e.touches) {
				mouseLocation = {
					x: e.touches[0].pageX,
					y: e.touches[0].pageY,
				};
			}

			// Normalize the angle to make 0 degrees the topmost point of the arc
			let angleDegrees =
				getAngleBetweenPoints(sliderCenter, mouseLocation) + 90;

			if (angleDegrees > 180) {
				angleDegrees = Math.max(angleDegrees - 360, -ANGLERANGE);
			} else if (angleDegrees < -ANGLERANGE) {
				angleDegrees = -ANGLERANGE;
			} else if (angleDegrees > ANGLERANGE) {
				angleDegrees = ANGLERANGE;
			}

			const value = (angleDegrees + ANGLERANGE) / (ANGLERANGE * 2);

			this.setState({
				innerValue: value,
			});

			if (this.props.onValueChange) {
				this.props.onValueChange(value);
			}
		}
	};

	render() {
		const { innerValue } = this.state;

		// If there is no value prop, use the innerValue so that the slider can
		// behave like an uncontrolled input
		const value =
			this.props.value == null
				? innerValue
				: constrainNumber(this.props.value, 0, 1);
		const { onValueChange, children, ...props } = this.props;
		const color = this.props.fillColor || '#FEC400';
		const background = this.props.background || '#707274';
		const valueAngle = ANGLERANGE * 2 * value - ANGLERANGE;

		return (
			<ArcSliderContainer {...props}>
				<StyledSVG
					style={{
						cursor: this.state.isDragging ? 'grabbing' : 'grab',
					}}
					innerRef={(elem: any) => (this.$slider = elem)}
					viewBox="0 0 600 600"
					onMouseDown={this.handleMouseDown}
					onTouchStart={this.handleMouseDown}
				>
					<defs>
						<linearGradient
							id="rendition-arc-slider-gradient"
							x1="0%"
							y1="0%"
							x2="100%"
							y2="0%"
						>
							<stop offset="0%" stopColor={darken(color)} />
							<stop offset="100%" stopColor={color} />
						</linearGradient>

						<filter
							xmlns="http://www.w3.org/2000/svg"
							id="rendition-arc-slider-dropshadow"
							height="130%"
						>
							<feGaussianBlur in="SourceAlpha" stdDeviation="1.25" />
							<feOffset dx="0" dy="0" result="offsetblur" />
							<feMerge>
								<feMergeNode />
								<feMergeNode in="SourceGraphic" />
							</feMerge>
						</filter>
					</defs>

					<path
						fill="none"
						stroke={background}
						strokeWidth="20.5"
						strokeLinecap="round"
						d={describeArc(300, 300, 200, -ANGLERANGE, ANGLERANGE)}
					/>

					<path
						fill="none"
						stroke="url(#rendition-arc-slider-gradient)"
						strokeWidth="22"
						strokeLinecap="round"
						d={describeArc(300, 300, 200, -ANGLERANGE, valueAngle)}
					/>

					<circle
						cx="300"
						cy="100"
						r="20"
						fill="white"
						filter="url(#rendition-arc-slider-dropshadow)"
						transform={`rotate(${valueAngle}, 300, 300)`}
					/>

					<circle
						cx="300"
						cy="100"
						r="14"
						fill={color}
						transform={`rotate(${valueAngle}, 300, 300)`}
					/>
				</StyledSVG>

				<LabelContainer>{children}</LabelContainer>
			</ArcSliderContainer>
		);
	}
}

export default ArcSlider;
