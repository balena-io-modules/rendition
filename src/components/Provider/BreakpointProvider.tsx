import debounce from 'lodash/debounce';
import * as React from 'react';
import { BreakpointContext } from '../../contexts/BreakpointContext';
export interface BreakpointProviderProps extends React.PropsWithChildren<{}> {
	breakpoints: number[];
	onBreakpointChange?: (breakpoint: number) => void;
}

export const BreakpointProvider: React.FC<BreakpointProviderProps> = ({
	breakpoints,
	children,
}) => {
	const [currentBreakpoint, setCurrentBreakpoint] = React.useState<number>(0);

	React.useEffect(() => {
		if (!breakpoints || !breakpoints.length) {
			return;
		}

		const handleResize = () => {
			const clientWidth = window.innerWidth;
			const breakpointIndex = breakpoints.findIndex(
				(breakpoint) => clientWidth < breakpoint,
			);
			const currentBreakpoint =
				breakpointIndex >= 0 ? breakpointIndex : breakpoints.length - 1;
			setCurrentBreakpoint(currentBreakpoint);
		};

		const debouncedResize = debounce(handleResize, 80);

		handleResize();

		window.addEventListener('resize', debouncedResize, { passive: false });

		return () => {
			window.removeEventListener('resize', debouncedResize);
		};
	}, [breakpoints]);

	return (
		<BreakpointContext.Provider value={{ currentBreakpoint }}>
			{children}
		</BreakpointContext.Provider>
	);
};

BreakpointProvider.displayName = 'BreakpointProvider';
