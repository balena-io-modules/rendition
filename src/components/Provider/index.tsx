import { Grommet } from 'grommet';
import merge from 'lodash/merge';
import * as React from 'react';
import styled from 'styled-components';
import { DefaultProps, Theme } from '../../common-types';
import defaultTheme from '../../theme';
import { px } from '../../utils';

const Base = styled(Grommet)`
	font-family: ${props => props.theme.font};
	font-size: ${props => px(props.theme.fontSizes[1])};

	h1,
	h2,
	h3,
	h4,
	h5,
	h6,
	button {
		font-family: ${props => props.theme.titleFont};
	}
`;

const Provider = ({ theme, ...props }: ThemedProvider) => {
	return <Base theme={merge(defaultTheme, theme)} {...props} />;
};

export interface ThemedProvider extends DefaultProps {
	theme?: Theme;
}

export default Provider;
