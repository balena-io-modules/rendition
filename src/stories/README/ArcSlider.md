# ArcSlider

A slider input that is displayed as an arc. This component will scale in size to
fit it's container. A label can be added by placing a child element inside this
component.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/stories/ArcSlider.js)

## Props

| Name          | Type      | Default   | Required   | Description                                                   |
| --------------|-----------|-----------|------------|-------------------------------------------------------------- |
| `onValueChange`    | `(value: number) => void` | - | - | A function that is called when the slider value changes, this will always be a value between 0 and 1 |
| `value`       | `number`  | -         | -          | A number between 0 and 1 that represents the progress         |
| `fillColor`   | `string`  | -         | -          | A CSS color string to use for the color of the "filled" part of the arc |
| `background`  | `string`  | -         | -          | A CSS color string to use for the color of the arc track      |

## Inheritance

The properties of the [`Box`][1] component are also available.

[1]: /?selectedKind=Box
