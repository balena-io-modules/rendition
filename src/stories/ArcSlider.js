import { storiesOf } from '@storybook/react'
import { withScreenshot } from 'storybook-chrome-screenshot'
import withReadme from 'storybook-readme/with-readme'
import * as React from 'react'
import styled from 'styled-components'
import { ArcSlider, Box, Provider } from '../'
import * as Readme from './README/ArcSlider.md'

const Container = styled(Box).attrs({
  m: 3
})`
  max-width: 450px;
`

class HOC extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      value: 0.6
    }

    this.handleChange = value => {
      this.setState({ value })
    }
  }

  render () {
    return (
      <ArcSlider value={this.state.value} onValueChange={this.handleChange}>
        <h2>{Math.round(this.state.value * 100)}%</h2>
      </ArcSlider>
    )
  }
}

storiesOf('Core/ArcSlider', module)
  .addDecorator(withReadme(Readme))
  .addDecorator(withScreenshot())
  .add('Standard', () => {
    return (
      <Provider>
        <Container>
          <ArcSlider />
        </Container>
      </Provider>
    )
  })
  .add('Fill color', () => {
    return (
      <Provider>
        <Container>
          <ArcSlider fillColor='#ff0000' />
        </Container>
      </Provider>
    )
  })
  .add('Background color', () => {
    return (
      <Provider>
        <Container>
          <ArcSlider background='#ff0000' />
        </Container>
      </Provider>
    )
  })
  .add('Label', () => {
    return (
      <Provider>
        <Container>
          <ArcSlider>
            <h2>A label</h2>
          </ArcSlider>
        </Container>
      </Provider>
    )
  })
  .add('Change handler', () => {
    return (
      <Provider>
        <Container>
          <HOC />
        </Container>
      </Provider>
    )
  })
