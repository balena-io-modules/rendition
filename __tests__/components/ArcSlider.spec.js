/* globals expect, describe, it */
import React from 'react'
import renderer from 'react-test-renderer'
import Provider from '../../src/components/Provider'
import ArcSlider from '../../src/components/ArcSlider'

describe('ArcSlider component', () => {
  it('should match the stored snapshot', () => {
    const component = renderer.create(
      <Provider>
        <ArcSlider />
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
