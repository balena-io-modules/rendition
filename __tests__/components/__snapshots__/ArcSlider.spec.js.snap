// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArcSlider component should match the stored snapshot 1`] = `
<div
  className="sc-bdVaJa jKopLr"
>
  <div
    className="sc-jTzLTM fRjxOf sc-gqjmRU bCjhLA sc-iwsKbI khHCQa sc-bwzfXH ihgQcP"
  >
    <svg
      className="sc-VigVT dFFOvp"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "cursor": "grab",
        }
      }
      viewBox="0 0 600 600"
    >
      <defs>
        <linearGradient
          id="rendition-arc-slider-gradient"
          x1="0%"
          x2="100%"
          y1="0%"
          y2="0%"
        >
          <stop
            offset="0%"
            stopColor="hsl(46.30000000000001, 100%, 39.8%)"
          />
          <stop
            offset="100%"
            stopColor="#FEC400"
          />
        </linearGradient>
        <filter
          height="130%"
          id="rendition-arc-slider-dropshadow"
          xmlns="http://www.w3.org/2000/svg"
        >
          <feGaussianBlur
            in="SourceAlpha"
            stdDeviation="1.25"
          />
          <feOffset
            dx="0"
            dy="0"
            result="offsetblur"
          />
          <feMerge>
            <feMergeNode />
            <feMergeNode
              in="SourceGraphic"
            />
          </feMerge>
        </filter>
      </defs>
      <path
        d="M 441.4213562373095 441.4213562373095 A 200 200 0 1 0 158.57864376269046 441.4213562373095"
        fill="none"
        stroke="#707274"
        strokeLinecap="round"
        strokeWidth="20.5"
      />
      <path
        d="M 158.57864376269046 441.4213562373095 A 200 200 0 0 0 158.57864376269046 441.4213562373095"
        fill="none"
        stroke="url(#rendition-arc-slider-gradient)"
        strokeLinecap="round"
        strokeWidth="22"
      />
      <circle
        cx="300"
        cy="100"
        fill="white"
        filter="url(#rendition-arc-slider-dropshadow)"
        r="20"
        transform="rotate(-135, 300, 300)"
      />
      <circle
        cx="300"
        cy="100"
        fill="#FEC400"
        r="14"
        transform="rotate(-135, 300, 300)"
      />
    </svg>
    <div
      className="sc-fjdhpX ioCTzx sc-gZMcBi cmZZgb sc-dnqmqq eheJhs sc-htpNat bhbuxU"
    />
  </div>
</div>
`;
